### ISSUE TITLE
OCI provider alias throws "bad configuration" error if settings are coming only from environment variables

### ISSUE THREAD

**Issue by kviktorman** [terraform-providers/terraform-provider-oci#1135]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version and Provider Version

```
$ terraform -v
Terraform v0.12.29
+ provider.oci v3.87.0
```

### Affected Resource(s)

* provider "oci"

### Terraform Configuration Files

```hcl

variable "oci_provider_tenancy_ocid" {}
variable "oci_provider_region" {}
variable "oci_provider_user_ocid" {}
variable "oci_provider_user_fingerprint" {}
variable "oci_provider_user_private_key_path" {}

provider "oci" {
  
  # remove the comment before the alias and run "terraform apply"
  # alias            = "region_europe"

  tenancy_ocid     = var.oci_provider_tenancy_ocid
  region           = var.oci_provider_region
  user_ocid        = var.oci_provider_user_ocid
  fingerprint      = var.oci_provider_user_fingerprint
  private_key_path = var.oci_provider_user_private_key_path
}

# Get a list of Availability Domains
data "oci_identity_availability_domains" "ads" {
  compartment_id = var.oci_provider_tenancy_ocid
}

# Output the result
output "show-ads" {
  value = data.oci_identity_availability_domains.ads.availability_domains
}
```

### Debug Output

<details>
<summary>Details</summary>
<p>

```
# ##############
# Init step succeeds
# ##############

$ terraform init

Initializing the backend...

Initializing provider plugins...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.oci: version = "~> 3.87"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

# #########################
# Alias commented out succeeds
# #########################

$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.oci_identity_availability_domains.ads: Refreshing state...

------------------------------------------------------------------------

No changes. Infrastructure is up-to-date.

This means that Terraform did not detect any differences between your
configuration and real physical resources that exist. As a result, no
actions need to be performed.

# #############
# Using alias on the provider fails 
# #############

$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


Error: can not create client, bad configuration: did not find a proper configuration for tenancy

  on <empty> line 0:
  (source code not available)

```

</p>
</details>  

### Expected Behavior

Reading and binding the provider settings from environment variables and aliasing the provider should not throw an error.

### Actual behaviour

I run into the following behaviour on Linux (don't know if it occurs on other OS): 

- When OCI terraform provider is declared without alias + settings are bind using environment variables, then terraform is capable to interact with the OCI platform
- When OCI terraform provider is declared and alias is used + settings are bind using environment variables, then terraform fails to interact with the OCI platform
- When OCI terraform provider is declared and alias is used + "SDK and CLI Configuration File" is present, but provider settings are declared coming from environment variables, then terraform is capable of interacting with the OCI platform.

### Steps to Reproduce

1. In order to reproduce do not use "SDK and CLI Configuration File", read the configuration only from environment variables and put an alias on the provider
2. run `terraform init`
3. run `terraform apply`, this will throw the error.

**Comment by rashik-bhasin** [comment]
Hi,
I tried to reproduce the error on Mac OS and on Linux machine but it's not reproducible. Assuming you have only one provider declared in your config. When you add an alias to the provider and if even one of the resources is not explicitly specifying the provider alias via provider = "oci.region_europe", then for these resources terraform creates a default provider other than your alias provider. The default provider loads configuration from /.oci or env variables. These variables may be supplying different values causing the failure. I would suggest either removing the alias in your config or every resource should specify the provider alias it wants to use. 
Also try sourcing your bash_profile(if you have your env variables set in it or any other file where you have your alias env variables) before running the terraform commands. It might be that your terraform configuration is picking different provider variables. 
Similar issue was discussed here: https://github.com/terraform-providers/terraform-provider-oci/issues/712 and the issue was related to resources picking default provider rather than alias provider.

**Comment by kviktorman** [comment]
Hi @rashik-bhasin,
sorry my mistake with the example, you are completely right, the components require provider specification when using aliases. 

I managed to find the root cause. I started to prepare an image to share with you, which led to find it.

The issue was, that we introduced recently a sub-module to a module and by accident we did not define the provider on that. The mistake was hidden, because we managed the terraform code from machines that have the ".oci" folder with default settings. So that specific module took secretly the defaults.

The error message in this case was not so user friendly/helpful, I'm glad that you managed to describe what might be the behaviour. Terraform was just putting its hands up and said that it has provider issues without giving any help which file/module/resource is involved. 

Thanks again to have a look and share root cause possibilities and solution ideas,
Viktor
